src/awkward/types/type.py

```diff
@@ -70,6 +71,14 @@ def _repr_args(self):
 
         return out
 
+    def is_equal_to(self, other, *, all_parameters: bool = False) -> bool:
+        return self._is_equal_to(other, all_parameters)
+
+    __eq__ = is_equal_to
+
+    def _is_equal_to(self, other, all_parameters: bool) -> bool:
+        raise ak._errors.wrap_error(NotImplementedError)
```

Comment:
The pattern here is:
- L2+ entrypoint with keyword-only parameters (anything with defaults should usually be kw-only)
- L3 protocol with positional no-default parameters (can be called on other types)

This means each type needs to implement the same guard (`isinstance(other, type(self))`), but I think the readability benefit is worth it. Existing checks here bounce between L2 and L3 functions.

Reply:
:+1: